### PR TITLE
stdio hardening + new mutex API

### PR DESCRIPTION
**Pull request by kilograham**
* Harden stdio_usb and stdio in general against deadlocks which could otherwise result from doing printfs from within IRQs
* Add a test for the above
* Add mutex_try_enter_block_until API.
* Make best_effort_wfe_or_timeout not use alarms if called from within IRQ

### CODE REVIEW

**Comment by kilograham** [comment]
fixes #1039

**Comment by kilograham** [comment]
fixes #1078 (probably)

File: src/rp2_common/pico_stdio_usb/stdio_usb.c [hunk 1]
```diff
@@ -127,19 +125,28 @@ static void stdio_usb_out_chars(const char *buf, int length) {
...
     int rc = PICO_ERROR_NO_DATA;
-    if (stdio_usb_connected() && tud_cdc_available()) {
-        int count = (int) tud_cdc_read(buf, (uint32_t) length);
-        rc =  count ? count : PICO_ERROR_NO_DATA;
-    } else {
-        // because our mutex use may starve out the background task, run tud_task here (we own the mutex)
-        tud_task();
-    }
-    mutex_exit(&stdio_usb_mutex);
+//    if (stdio_usb_connected() && tud_cdc_available()) {
```
Comment (peterharperuk) [inline src/rp2_common/pico_stdio_usb/stdio_usb.c]: did you mean to leave this code commented out

Comment (peterharperuk) [reply src/rp2_common/pico_stdio_usb/stdio_usb.c]: ah - I see the second commit now

Comment (kilograham) [reply src/rp2_common/pico_stdio_usb/stdio_usb.c]: no, but i had already fixed that!

**Review by peterharperuk — APPROVED** [review]

**Pull request merged**